tensorflow/compiler/tf2tensorrt/convert/ops/layer_utils.h

```diff
@@ -0,0 +1,365 @@
+/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#ifndef TENSORFLOW_COMPILER_TF2TENSORRT_CONVERT_OPS_LAYER_UTILS_H_
+#define TENSORFLOW_COMPILER_TF2TENSORRT_CONVERT_OPS_LAYER_UTILS_H_
+#if GOOGLE_CUDA && GOOGLE_TENSORRT
+
+#include <type_traits>
+
+#include "tensorflow/compiler/tf2tensorrt/convert/convert_nodes.h"
+#include "tensorflow/compiler/tf2tensorrt/convert/utils.h"
+#include "tensorflow/core/framework/types.h"
+#include "tensorflow/core/platform/statusor.h"
+#include "third_party/tensorrt/NvInfer.h"
+#include "third_party/tensorrt/NvInferRuntimeCommon.h"
+
+namespace tensorflow {
+namespace tensorrt {
+
+namespace convert {
+
+// Simplifies the creation of a TensorRT network. The user provides a
+// INetworkDefinition pointer during construction. They can then add operations
+// to the network through the provided functions. Each function returns a struct
+// which contains the symbolic result of the operation (ITensor pointer) as well
+// as a pointer to the last TensorRT ILayer created. Some operations may create
+// multiple layers in order to accomplish the desired result (e.g. Sign).
+class TRTNetworkBuilder {
+ public:
+  TRTNetworkBuilder(nvinfer1::INetworkDefinition* network,
+                    TrtWeightStore* weight_store)
+      : network_(network), weight_store_(weight_store) {}
+
+  // Adds an Add operation to the network.
+  StatusOr<nvinfer1::IElementWiseLayer*> Add(nvinfer1::ITensor* lhs,
+                                             nvinfer1::ITensor* rhs) noexcept {
+    TRT_EXPECT(network_);
```

Comment:
Can we avoid TRT_EXPECT(network_) in all the functions that use it by making
"network_" "nvinfer1::INetworkDefinition* const network_" and only check it in the constructor?

Reply:
U don't think so, doing this doesn't provide the same guarantee of catching misuse.